dimod/include/dimod/expression.h

```diff
@@ -0,0 +1,587 @@
+// Copyright 2022 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#pragma once
+
+#include <limits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "dimod/abc.h"
+#include "dimod/vartypes.h"
+
+namespace dimod {
+
+// forward declaration
+template <class Bias, class Index>
+class ConstrainedQuadraticModel;
+
+template <class Bias, class Index = int>
+class Expression : public abc::QuadraticModelBase<Bias, Index> {
+ public:
+    /// The type of the base class.
+    using base_type = abc::QuadraticModelBase<Bias, Index>;
+
+    /// The first template parameter (Bias).
+    using bias_type = Bias;
+
+    /// The second template parameter (Index).
+    using index_type = Index;
+
+    /// Unsigned integral type that can represent non-negative values.
+    using size_type = std::size_t;
+
+    using parent_type = ConstrainedQuadraticModel<bias_type, index_type>;
+
+    /// @private  <- don't doc
+    class ConstQuadraticIterator {
+     public:
+        using value_type = abc::TwoVarTerm<bias_type, index_type>;
+        using pointer = const value_type*;
+        using reference = const value_type&;
+        using difference_type = std::ptrdiff_t;
+        using iterator_category = std::forward_iterator_tag;
+
+        ConstQuadraticIterator() : expression_(nullptr), it_(), end_(), term_(0) {}
+        ConstQuadraticIterator(const Expression* expression,
+                               const abc::ConstQuadraticIterator<bias_type, index_type>& it,
+                               const abc::ConstQuadraticIterator<bias_type, index_type>& end)
+                : expression_(expression), it_(it), end_(end), term_(0) {
+            if (it_ != end_) {
+                term_ = value_type(expression_->variables()[it_->u],
```

Comment:
Could write a short function to create a term from interator and use it here and below.